lib/graphql/schema/input_object.rb

```diff
@@ -184,21 +184,25 @@ def validate_non_null_input(input, ctx)
             end
           end
 
-          visible_arguments_map = warden.arguments(self).reduce({}) { |m, f| m[f.name] = f; m}
-
-          # Items in the input that are unexpected
-          input.each do |name, value|
-            if visible_arguments_map[name].nil?
-              result.add_problem("Field is not defined on #{self.graphql_name}", [name])
+          # Inject missing required arguments
+          required_inputs = self.arguments.reduce({}) do |m, (argument_name, argument)|
+            if !input.key?(argument_name) && argument.type.non_null? && warden.get_argument(self, argument_name)
+              m[argument_name] = nil
             end
+
+            m
           end
 
-          # Items in the input that are expected, but have invalid values
-          visible_arguments_map.map do |name, argument|
-            argument_result = argument.type.validate_input(input[name], ctx)
-            if !argument_result.valid?
-              result.merge_result!(name, argument_result)
+          input.to_h.merge(required_inputs).each do |argument_name, value|
```

Comment:
Trying this out in Shopify, `input` is sometimes a `ActionController::Parameters` instance, e.g. `<ActionController::Parameters {"barcode"=>"53941047", "id"=>"gid://shopify/ProductVariant/30322695"} permitted: false>`.

This causes `input.to_h` to raise `ActionController::UnfilteredParameters: unable to convert unpermitted parameters to hash`.

I notice that above this line we have this block:

```ruby
begin
  input.to_h
rescue
  begin
    # Handle ActionController::Parameters:
    input.to_unsafe_h
  rescue
    # We're not sure it'll act like a hash, so reject it:
    result.add_problem(INVALID_OBJECT_MESSAGE % { object: JSON.generate(input, quirks_mode: true) })
    return result
  end
end
```

Which doesn't actually convert `input` if it's a `ActionController::Parameters`. Should it do so?

Elsewhere in this gem, I see [handling for this same scenario](https://github.com/rmosolgo/graphql-ruby/blob/c67be42/lib/graphql/language/sanitized_printer.rb#L154-L160), checking if `value.respond_to?(:to_unsafe_h)`.

Reply:
Previously, it required that `.each` yielded key-value pairs. I think making a hash out of it is fine, if you have to, too 🤷